### ISSUE TITLE
[Feature request] Support texlab language server

### ISSUE THREAD

**Issue by krishnakumarg1984** [lervag/vimtex#1371]
I had lots of fun using `vimtex` for writing a paper and my PhD thesis :). Thank you for developing and maintaining this modern plugin.

It is 2019 and Language servers are here! VSCode seems to be having a runaway success due to sharing common origins at the same company that formulated the nicely documented, open & editor-agnostic [language server protocol (LSP)](https://langserver.org/). 

### Servers
The idea is that all aspects that are specific to a particular programming language are left to a dedicated language server. A compatible client (i.e. a text editor) can simply ask questions to the server and populate its UI with the response received. Various common tasks such as `completions`, `hover-text`, and `go-to-definitions` are thus offloaded to the server. Though not all language servers are mature yet, the well-documented, open specification means that these get continuous refinements from the community.  The official set of language servers for various programming languages is available [here](https://microsoft.github.io/language-server-protocol/implementors/servers/).

### Clients (for vim)
For `vim`, the language server capability (i.e. the client functionality that understands LSP lingo) is provided by a couple of plugins, notably `[languageclient-neovim](https://github.com/autozimu/LanguageClient-neovim)` (python) and `[vim-lsp](https://github.com/prabirshrestha/vim-lsp)' (vimscript).  A [pull-request](https://github.com/neovim/neovim/pull/6856) to bring in native LSP functionality to `neovim` is fairly mature & scheduled to land for v0.5. 

### LaTeX-specific information
The officially supported language server for LaTeX is [TexLab](https://texlab.netlify.com/) which also handles BibTeX.  I have had a positive experience using it in VSCode for which has a client plugin bearing the same name (TexLab). 

### Question
Can `vimtex` provide support for (i.e. play nicely with) TexLab? Already the modular codebase & design goals of vimtex instruct to use dedicated plugins for tasks such as completion, folding and reference doc. Instead of using a variety of plugins, using a language server might be a really good idea. 

I am not sure what `vimtex` needs to do to support LSP. At the least, this possibility should be mentioned in the documentation (but only after testing).

**Comment by lervag** [comment]
> Thank you for developing and maintaining this modern plugin.

I'm happy you like it! :)

> Can vimtex provide support for (i.e. play nicely with) TexLab?

I do not quite see _how_. That is, `vimtex` is not a language server client. My initial thought is that it does not make sense for `vimtex` to adapt/utilize TexLab because there already exists mature LSP clients for Vim that should be able to support TexLab already.

I agree that it may be relevant to mention TexLab and perhaps some information about how to use it with Vim in the docs. But I have not used it myself and I have quite limited time. As such, I will gladly review and accept PRs for this, but I will probably not do anything myself in any time soon.

**Comment by krishnakumarg1984** [comment]
@lervag  Okay here are my thoughts. 

How about vimtex provides a user option `vimtex_override_with_lsp` which is default set to `0` but the user can explicitly set it to `1` if using LSP.

Now, this option is just a shortcut for disabling the features set by the LSP (to avoid conflict).

As far as I am aware, LSPs won't ever provide 80-90% of the stuff that vimtex currently handles (correctly handling tex root file, folding, setting tex engines and options, LaTeX specific text objects etc.), so I see that using vimtex is an absolute must i.e LSP is not a replacement for vimtex.

IMHO, an LSP complements vimtex's strengths with only a small overlap. My request here is to implement a design whereby vimtex acts merely like a "passthrough" to the LSP for majority of the features. Being a modular codebase, I think this isn't as difficult as it initially appears.

**Comment by lervag** [comment]
You raise a fair point, and I can agree to adding such an option. But as said, I do not have the time to investigate this in detail myself. If anyone would be able/willing to provide a reproducible example of how to setup Vim with vimtex, an LSP plugin and the TexLab server, and with that a way to reproduce overlap issues, then I could implement a robust version if this option.

Note that vimtex does already provide a lot of `g:vimtex_..._enabled` options that allows to disable various parts.

**Comment by teto** [comment]
Looks like texlab is being rewritten in rust so let's wait and see what features get available as it's very immature. As LSP providers improve, vimtex could lose some weight :)

**Comment by lervag** [comment]
If (and when) other plugins/providers are made available that renders some of vimtexs features deprecated, then I'm just glad. I would be happy to make vimtex work well alongside LSP plugins.

In the meantime, I'm closing this as there is nothing to do right now.

**Comment by clason** [comment]
Out of curiosity, I spent some time yesterday playing around with this. **TL;DR:** it works, offers some neat functions that vimtex doesn't, doesn't do many things that vimtex does, and where there is overlap, vimtex does it better. Could also use a bit more polish (which it is getting though). 

If anyone else wants to play with it (or is just interested in what it does), here's more details:

## Setup

First, download https://github.com/latex-lsp/texlab/releases/download/v0.4.2/texlab.jar and put it somewhere. (I've tried compiling the Rust version, but this didn't work yet -- not that it was meant to.)

Next, install a LSP client. I used [prabirshrestha/vim-lsp](https://github.com/prabirshrestha/vim-lsp) for this (there's also [natebosch/vim-lsc](https://github.com/natebosch/vim-lsc), but that worked less well for me). It's a pure VimL implementation, so all you have to do is to add the following to your .vimrc:

```viml
Plug 'prabirshrestha/async.vim'
Plug 'prabirshrestha/vim-lsp'
au User lsp_setup call lsp#register_server({
        \ 'name': 'texlab',   
        \ 'cmd': {server_info->['java','-jar','PATH_TO_texlab.jar']},
        \ 'whitelist': ['bib','tex'],
        \ })
```

The language server should now start automatically as soon as you open a tex or bib file.

(Alternatives would be neoclide/coc.nvim and autozimu/LanguageClient, which are more full-featured but need a binary install. If anybody has tested this and gotten better results than I report below, I'd be interested to hear about it.)

## LSP-only features

You can then do the following things, which may or may not be useful:
1. Go to definition (`LspDefinition`) of labels and citations (and commands? that didn't work for me) under the cursor.
2. Conversely, you can list all occurrences (`LspReferences`) of a label under the cursor in the quickfix list. This also works in a bib file. (You can also step through occurrences with `LspNextReference`.)
3. Get a preview of citations and equations (`LspHover`) under the cursor. (Citations give you a formatted text, equations a markdown formatted image -- not really useful for a text-based editor... but might be neat in a GUI that has builtin markdown preview. )
4. Rename (`LspRename`) *all* occurrences of a label or citations under the cursor (think: `\eqref{eq1}` to `\eqref{eq:boundedness}` or `\cite{c1}` to `\cite{Clason:2011}` -- yes, in the bib file too)
5. It also does linting (using virtual text if available) for bib (but apparently not tex) files.
6. Highlight groups for errors, warnings, informations, hints (for bibtex only, I think). It also highlights all references of a symbol under the cursor (seems more useful for actual programming languages, though).


## LSP and vimtex features

The only real overlap that I noticed was completion; you can hook the LSP up to omnicomplete via `:set omnifunc=lsp#complete` (it's not enabled by default). It seems quite a bit faster than the native VimL from vimtex, but this may be due to the problem with project-wide parsing I mentioned; the pum entries are also not styled like vimtex's -- but there's a hook that allows doing that in principle, and this I would see as the single feature that vimtex could support a TeX lsp with. It seems specific to each LSP plugin, though, so a solution for vim-lsp would probably not work for, e.g., vim-lsc.

## What doesn't work (yet)

1. Most notably, automatic project detection, which should identify all TeX files belonging to a document; I only ever got results from the local file.
2. You also need to use the extension for the bibfile (`\bibiliography{bibfile.bib}`) even for BibTeX to get bibliography-related features to work; 
3. There were also problems with multiple references per command, `\cite{A,B}`. 
4. Cleveref support.

(1. maybe I did something wrong; hints welcomed; 2. and 3. look like bugs with either the language server or the client and should be reported; 4. is being worked on at the moment.)

**Comment by krishnakumarg1984** [comment]
@clason That's awesome! Have you tried to see if it plays well with [coc-vimtex](https://github.com/neoclide/coc-vimtex), which is an LSP client?

**Comment by clason** [comment]
No; as I wrote, I don't want to mess with binary installs. I only tried pure VimL clients. (And some of these problems also occur with the vscode extension, so they are definitely texlab bugs.)

**Comment by lervag** [comment]
I'm happy to see more information about this. I would be happy to document this somehow. If any of you would be willing and able to suggest something, that would be very good. I'm thinking this might earn a spot at the top of `vimtex-non-features`, at least a small text there, perhaps with a link to a section where we can guide users on how to make the most of a combination of vimtex and texlab?

**Comment by clason** [comment]
I was planning to offer to write up my notes properly as a section for the Wiki, once texlab has a full release (probably very soon) and I get the last kinks in the setup ironed out (TeX is not a standard language for a server, and vim is not a standard editor for a client, so there are some edge cases both ways. Also, I can't brain.)

**Comment by lervag** [comment]
Great, and thanks!

> Also, I can't brain.

I can relate to this too often these days...

**Comment by clason** [comment]
Thinking a bit more about this, LSP integration may not have to be a `non-feature` (although I can definitely see why it may be, see below). What a language server provides is a structured, language-aware, access to the code you're editing. For example, it knows which parts of the code (in its language) are meaningful "symbols" -- for TeX, mostly labels (and bib keys) and keeps track of where they are. Vimtex could leverage this more powerful (and faster!) access for features which it currently uses regex grepping and aux files for -- such as autocompletion or table of contents. This means outsourcing "lower-level" parts of vimtex and concentrating on how to present and manipulate the results, beyond the standard Actions defined by the generic language server protocol. Neovim is currently working on implementing low-level LSP integration that could be used for that (so no extra plugins necessary).

Clearly, that would be a massive refactor, and the reliance on an external binary for important functionality might be distasteful (it certainly goes against the vim principle of ubiquity and platform-independent scripting). So unless you need a new hobby -- "vimtex 2.0" -- it would probably need to be a fork -- "nvimtex", if it uses the new neovim hooks -- and hence "someone elses problem" :)

**Comment by lervag** [comment]
> This means outsourcing "lower-level" parts of vimtex and concentrating on how to present and manipulate the results, beyond the standard Actions defined by the generic language server protocol.

I thought this was already well provided by the language server plugins such as `vim-lsp`, ALE, `coc-nvim`, `LanguageServer-nvim`, and so on. As such, it feels wrong to implement it as a dedicated LaTeX interface. I feel it should be more worthwhile to instead try and adapt vimtex to a workflow where some features are handled by the language server client. In a (probably quite distant) future, one could foresee that vimtex is greatly simplified in the sense that features like completion, ToC, etc, are handled by the language server.

> (Regarding "nvimtex": Neovim is also working on adding multiprocessing, and the first pull request reports an impressive speedup of 20x for vimgrep. Might be worth looking into once it lands!)

Thanks, this is very interesting information! I am happy to see that there are more enhancements in the pipeline for neovim.

**Comment by clason** [comment]
Well, the point of LSP (the protocol) is that it defines standard actions that *all* language servers (can) provide and *all* clients (can) trigger -- independent of the language. There is of course some room for adaption here -- for example, the "symbols" for python (variables, functions) are different than the ones for TeX (labels, bib keys). But the actions are fixed: rename, list, preview, etc. Many things that vimtex does do not easily fit into that mold (for example, the table of contents functionality is much more nuanced than simply "list all references"). And things like "change surrounding environment" or "toggle star" are far outside the protocol. Nevertheless, the basic infrastructure of the language server (compiled syntax-aware document parsing) might be leveraged in place of interpreted VimL regexp parsing for this if it is exposed to a higher level, which the neovim pull request seems to be allowing (if I interpret it correctly). Imagine what you could do if you had access to a list of all labels with their location in *all* files in the current project without having to manually parse the TeX (and aux!) files yourself. 

Of course, at this time, it's just pie-in-the-sky futurist hot air...

(And if you think that's cool, you should look at the tree-sitter pull-requests -- direct access to the syntax tree of a document generated by a compiled parser :))

**Comment by lervag** [comment]
> Imagine what you could do if you had access to a list of all labels with their location in all files in the current project without having to manually parse the TeX (and aux!) files yourself.

Good point, that would of course be very nice!

I'm also looking forward to the tree-sitter stuff, although I currently only know about it from reading on reddit. But it sounds very useful!

**Comment by clason** [comment]
Of course, the first thing I did was google "tree-sitter latex" and was happy to find https://github.com/yitzchak/tree-sitter-latex :)

**Comment by astier** [comment]
Would it be a possibility to turn vimtex to a language-server?

**Comment by clason** [comment]
@astier That doesn't really make sense -- the whole point of a language server is that it's independent of the editor.

Are there any specific features from texlab that you think vimtex should implement directly?

(@lervag If you're using coc, there's now https://github.com/fannheyward/coc-texlab that should make it fairly easy to try out in case you're curious. Watch out for issues with non-ASCII characters -- LSP is based on UTF-16, and that trips up for example vim-lsp. If it works for you, let us know!)

**Comment by astier** [comment]
Just forget my idea.

**Comment by clason** [comment]
@astier I did not mean to be dismissive, just asking for more specific details about your idea.

**Comment by astier** [comment]
No worries. I did not perceive it as dismissive. I just had a thought but it doesn't make sense.

**Comment by motorto** [comment]
@clason sorry to dig and old thread but vimtex is extraordinary as an plugin with auto-completion , previously I have used coc nvim , but have ever since moved to [vim-lsc](https://github.com/natebosch/vim-lsc), is there any plan to support or is already supported ? 
On help page (vimtex-complete) there are many lsc client supported but vim-lsc is not one .

**Comment by clason** [comment]
@motorto Vimtex is not a language server (and coc is not exactly an lsp client), so you're looking for the wrong thing, I'm afraid. What you want is a general-purpose completion plugin like deoplete, ncm2, or asyncomplete.

(I personally am not using autocomplete for vimtex and just trigger it manually.)

**Comment by motorto** [comment]
@clason. You are right ! I am aware that coc.nvim is more than lsc client.

Unfortunately texlab server is not on pair to vimtex at the moment so I am moving back to coc nvim to use coc-vimtex.

**Comment by clason** [comment]
Why can't you use both?

**Comment by motorto** [comment]
Vimtex and texlab ? 

Well to be honest I am having trouble configuring the texlab, their
documentation is not great (at least from my point of view).

Texlab gives me the autocompletition that I want Packages for example (I
couldn't make texlab do that).

If its the move from vim-lsc to coc-nvim , well then coc-nvim does
everything that vim-lsc does so , it's redundant having both :)

**Comment by krishnakumarg1984** [comment]
@lervag  I think now is a great time to revisit this. LSPs have become incredibly popular and are frankly quite awesome (I have started using `ccls`/`clangd` for my C++ mini-projects, and Gosh, it lends an order of magnitude improvement in quality of life!

### Why is this the best time to revisit this issue?

[Neovim 0.5](https://neovim.io/news/) has been just released. In my view, this is a game changer, with its built-in LSP integration, and the possibility of allowing lua plugins, tree-sitter integration etc.

In fact, @clason and I were briefly discussing an issue this morning about tree-sitter diff-ing in neovim. Frankly, the availability of some high quality plugins leveraging the `lua` support that neovim 0.5 provides has made previously unimaginable things happen.

I think that `vimtex` can perhaps look at offloading certain functionality to `texlab`, and focus on the core parts in which it excels.  `vimtex` may provide a user-meta option to turn on LSP (which in-turn can be an alias for turning off syntax highlighting, formatexpr, etc etc.). For those without access to an LSP client and server (or those who don't wish to use one), then we fall back to the complete functionality and goodness that `vimtex` offers. 

Of course, we need to think this through if there is further interest in this issue.

**Comment by clason** [comment]
I have been summoned!

Frankly, the arguments are still the same -- VimTeX targets a much broader audience than just neovim 0.5+ folks, and the overlap with LSP is not _that_ large, considering what VimTeX offers. (The argument is different from the other side, but again, texlab is primarily -- although happily not exclusively -- aimed at VS Code users.) And as the [release newsletter](https://neovim.io/news/) hopefully makes clear, tree-sitter is _not_ ready yet for production use and cannot replace the extensive (in particular, package-specific!) syntax support of VimTeX.

And there is nothing preventing you from using both texlab and VimTeX (and tree-sitter, if you're feeling brave) in parallel; I've been doing so for more than a year now.

Is there an opportunity for an "NvimTeX" built on top of LSP, tree-sitter, and whatever other goodies Neovim offers (Telescope? LuaSnip?) Absolutely.  But It Is Not This Day.

**Comment by clason** [comment]
(And I know you don't intend it that way, but this suggestion has an aftertaste of "now that LSP is out, you can stop trying and leave this to the experts" -- you're essentially suggesting that @lervag abandon years of work and start from scratch!)

**Comment by krishnakumarg1984** [comment]
@clason Absolutely not. Karl is one of my favourite open-source developers, and has been an absolutely awesome human being to interact with (I have talked with him once). `vimtex` is a gem of a plugin, and I used it to write 4 papers and a PhD thesis, and I am ever grateful for it. 

My point was to provide a user-flag to selectively disable certain groups of configs that can be better done by an external tool. For instance, it is well-known that `latexformat.pl` is a better customised formatting tool for tex files even though vimtex's internal formatter is no slouch.

I advocate co-existence of vimtex and LSP. My point was to have users a bit of control through user setting flag, and if no LSP is found, then we fall back to the complete set of features that `vimtex` provides.

**Comment by clason** [comment]
It's still not clear what exactly you want changed, then; there's zero conflict between VimTeX and texlab (and tree-sitter), they happily coexist already.

**Comment by lervag** [comment]
This has already had many words, so I'll try to first make some general responses and comments, then I'll close with coments to the more concrete suggestions.

First, I really do not mind these discussions. I don't want to end up maintaining a plugin that "lives in the stoneage", so I need to be aware of what is happening around me. But my time is quite limited, so it is very good that the community helps me get informed! Tree-sitter and LSPs are huge, and at some point in time I believe it _will_ impact how I shape VimTeX!

### General

> LSPs have become incredibly popular and are frankly quite awesome (I have started using `ccls`/`clangd` for my C++ mini-projects, and Gosh, it lends an order of magnitude improvement in quality of life!

I strongly agree on the usefulness of LSPs!

> @teto texlab has indeed been re-written in pure rust, and have been making regular releases. Indeed, the [project repo](https://github.com/latex-lsp/texlab) is super active with frequent commits.

I'm glad to hear it! I still don't really use texlab much.

> [Neovim 0.5](https://neovim.io/news/) has been just released. In my view, this is a game changer, with its built-in LSP integration, and the possibility of allowing lua plugins, tree-sitter integration etc.

I agree this is all good! :)

> I think that `vimtex` can perhaps look at offloading certain functionality to `texlab`

I would not really mind offloading things, but it has to be done very carefully. I personally don't see any good reason to look into this right now.

As @clason writes (and I could not write it better myself):

> Frankly, the arguments against are still the same -- VimTeX targets a much broader audience than just neovim 0.5+ folks, and the overlap with LSP is not _that_ large, considering what VimTeX offers.
> ...
> And there is nothing preventing you from using both texlab and VimTeX (and tree-sitter, if you're feeling brave) in parallel

### Concrete

> `vimtex` may provide a user-meta option to turn on LSP (which in-turn can be an alias for turning off syntax highlighting, formatexpr, etc etc.). For those without access to an LSP client and server (or those who don't wish to use one), then we fall back to the complete functionality and goodness that `vimtex` offers.

I don't really see a very simple way to do this. VimTeX would have to properly detect the settings of the LSP. Some people may want some parts of the LSP, others may want other parts. So, it seems to me to be much better to simply allow people to properly disable parts of VimTeX that they want and let them customize their environment manually.

> I advocate co-existence of vimtex and LSP. My point was to have users a bit of control through user setting flag, and if no LSP is found, then we fall back to the complete set of features that vimtex provides.

My last point stands: I don't really see any clear/simple way for VimTeX to
understand what the user wants. Like:

* How do I detect if a user is running texlab? (Does it even start before VimTeX is initalized? Does this depend on the user config?)
* How do I detect which features a user wants from texlab?

**HOWEVER**: I do see how it could be beneficial to the community to have a better documentation on how to use and configure VimTeX alongside texlab. Currently, there is only a very small section that explains what texlab is. I would be both grateful and happy to get help from someone (perhaps @clason?) to write a better guide on how to make these plugins coexist; e.g. suggested configuration and describe where there is an overlap and so on. Since I still don't really use texlab, I don't know enough to write something like this myself.

**Comment by clason** [comment]
Well, my 

* LSP config is here: https://gist.github.com/clason/04b78ccbf49c4e4a3301b18927a83329
* Vimtex config is here: https://gist.github.com/clason/12f884907a80baec6bf9c18a467ff300
* mappings are here: https://gist.github.com/clason/3156fd207b16afc43cb2d5c911d35dc9

**Comment by lervag** [comment]
How does the texlab completion work compared to VimTeX?

**Comment by clason** [comment]
Cite: (texlab first, then VimTeX)

<img width="355" alt="Screenshot 2021-07-12 at 22 37 05" src="https://user-images.githubusercontent.com/2361214/125352541-b00e6b80-e361-11eb-97ac-c55cfb8ec985.png">

<img width="766" alt="Screenshot 2021-07-12 at 22 37 33" src="https://user-images.githubusercontent.com/2361214/125352598-c1577800-e361-11eb-8f9c-c36a5413e404.png">

Labels: (same)
<img width="766" alt="Screenshot 2021-07-12 at 22 38 06" src="https://user-images.githubusercontent.com/2361214/125352653-d502de80-e361-11eb-9f71-c664d6d6f508.png">

<img width="766" alt="Screenshot 2021-07-12 at 22 38 24" src="https://user-images.githubusercontent.com/2361214/125352687-df24dd00-e361-11eb-9763-a719cb46d29c.png">

(texlab is _much_ quicker, at least on a cold cache)

**Comment by lervag** [comment]
Thanks!

**Comment by clason** [comment]
(you could also link neovim's builtin LSP client: https://github.com/neovim/nvim-lspconfig/blob/master/CONFIG.md#texlab ;)

**Comment by krishnakumarg1984** [comment]
@clason Thank you for the useful comments. Can you please clarify what you meant by `Here, VimTeX can of course match on menu, which texlab can't.`

@lervag Thanks a lot for providing detailed explanatory answers. I think we can rest this case for now, bearing in mind the awareness that the neovim ecosystem is undergoing some big changes.

**Comment by 00sapo** [comment]
@clason this is an unfair comparison: lsp clients usually provide a preview when a completion entry is selected

**Comment by clason** [comment]
Sorry, I stand by this, which was a specific and not a general point: VimTeX offers a _better_ preview than texlab. (I use both -- and they're both great at what they do -- so I know.)

**Comment by krishnakumarg1984** [comment]
Both `texlab` & `vimtex` are being actively developed. The `texlab` LSP server is a rust application (which, depending on the camp you are in, might either come across as performant & type-safe, or as a mere trend-following/hype of the current era). 

Putting texlab to the side and thinking about the LSP ecosystem in general, the protocol is evolving well, and the information exchange between a client and server is a rather elegant JSON RPC dataframe. The simplicity of the protocol is encouraging more developers to contribute to various LSP servers, and generally I have seen better/faster completions, definition support etc in python/C++. 

Why did I say all these? I have truly benefited from VimTeX, and I agree with clason that it is _currently_ superior to texlab. It is imperative that, with the 7.5 years of hard work put into vimtex to make it so brilliant, it needs to remain relevant in the next decade. VimTeX has been tested against so many corner cases, and lervag writes up test cases with each PR to ensure that regressions are rare. 

I think, a visionary/prudent approach for VimTeX to thrive would be to: study exactly what features overlap with texlab (and the LSP protocol), and provide user options to switch to texlab if installed.

**Comment by clason** [comment]
I think VimTeX is thriving very well by doing what it is doing currently; I don't see any grounds on thinking that this will change in the near future. And at the risk of sounding like a broken record: VimTeX and Texlab coexist _very_ well already; if there is an overlap of functionality, you can set up your mappings to use one over the other as you prefer. There is _nothing_ VimTeX needs to do to support Texlab!

In general: such "philosophical" comments are rarely helpful; it's much better to focus on single concrete issues and tackle them one at a time.

**Comment by krishnakumarg1984** [comment]
I disagree that philosophical comments are rarely helpful. The Agile manifesto, which is now the backbone of modern software development, was an entirely philosophical in nature!

I think major open source projects _welcome_ discussions about user experiences, and have RFCs regarding its future direction. It would provide a more inclusive environment which will empower less technical people to contribute to the project.

**Comment by lervag** [comment]
@krishnakumarg1984 Thanks; I actually do appreciate philosophical comments and thoughts, and it is useful to get this kind of input. I agree with @clason's sentiment that concrete issues are in general much more helpful, but sometimes it is also very good to take a step back and consider things in a larger perspective.

Some thoughts to your concerns:

> study exactly what features overlap with texlab (and the LSP protocol), and provide user options to switch to texlab if installed.

Well, I'll try, or at least I'll accept PRs with updates to the docs and similar.

> use the dissection study from above to integrate beautifully with texlab, and provide user instructions in the documentation on how to integrate the two (It looks like clason has already integrated these two and his config could be a good starting point).

In this I agree with @clason:

> I think VimTeX is thriving very well by doing what it is doing currently; I don't see any grounds on thinking that this will change in the near future. And at the risk of sounding like a broken record: VimTeX and Texlab coexist _very_ well already; if there is an overlap of functionality, you can set up your mappings to use one over the other as you prefer. There is _nothing_ VimTeX needs to do to support Texlab!

However, I will admit that this is also somewhat "lucky". I currently (and still) do not use texlab myself, and as such, I do not experience problems that I thus fix for my own needs. So if there are concrete shortcomings, these should be raised by users. And in this particular case; ...

> In general: such "philosophical" comments are rarely helpful; it's much better to focus on single concrete issues and tackle them one at a time.

... as mentioned, I agree with @clason. The more "day to day" improvement and maintenance of VimTeX is very much dependent on users raising specific and concrete issues that I and (as I'm very glad to see!) others can address.

> The modus operandi would be to try this out in a feature branch, and announce `vimtex 3.0` series with LSP support/integration.

I do not fully understand what this integration would mean, specifically.

> Ignoring a fledgling entrant with potential has historically led to the downfall of so much software that were considered unsurmountable in their heydays.

I am not _knowingly_ ignoring texlab. But I am not able to spend as much time with VimTeX as I need to get on top of everything (there are still ~10 open issues and a couple of PRs, and some of these will correspond to major efforts). I don't think you mean it that way, of course.

> Can we at least provide an objective comparison of each overlapping capability in the documentation?

I think this sounds like a reasonable request. However, I would not call it comparison, instead I would call it a documentation of "integration" in the sense of explaining a "best practise" manner to combine VimTeX and texlab. I would welcome a PR for this. texlab is currently described in `:help vimtex-nf-lsp`, so a good starting point would be to update this text and more precisely describe the overlap.

In this sense, I think the following are things that would be useful for me to know in order to understand "the future of VimTeX":

* Which features of texlab are directly overlapping with VimTeX, now and in the future?
* Which of these should I assume will be "better" served by texlab as it progresses? All?
* In a similar vein: Will Tree-sitter syntax support for LaTeX eventually become superior? I would not be surprised if it does; if so, how fast is this development going?
* The above, combined and inversed: If we assume that texlab and Tree-sitter were "optimal", which VimTeX features would still be relevant?

**Comment by krishnakumarg1984** [comment]
@lervag Thanks for your receptive and detailed response. I agree to everything you have said here.  Just one minor point

>texlab is currently described in :help vimtex-nf-lsp, so a good starting point would be to update this text and more precisely describe the overlap.

The __non-features__ wording might be perceived as sending out a stronger negative signal. I know its pure semantics, but it appears as saying "No. This particular aspect is __not__ vimtex's job. If you want feature X, read this section and DIY". 

Instead, I think the section could be titled "Integration with other relevant technologies" or "Interoperation with other plugins" or something which conveys a more inclusive/welcoming tone (reflecting the openness of your response here). In this section, documenting an example set of keymaps and configs explaining how to co-operate with texlab shall enrich the user experience imo.

PS: Treesitter for LaTeX is currently nascent, but I suspect texlab development may accelerate treesitter as one feeds on the other.

**Comment by clason** [comment]
> PS: Treesitter for LaTeX is currently nascent, but I suspect texlab development may accelerate treesitter as one feeds on the other.

I didn't want to comment anymore but feel like I must correct you here: the `tree-sitter` parsers are _not_ used by texlab -- the developers investigated that at some point but in the end decided to use their own, special-purpose, parser. (There is a lesson here.) They polished up and released their parser on my request so that it may be used with nvim-treesitter. They have _not_ committed to maintaining/extending it in any way.

**Comment by krishnakumarg1984** [comment]
Thanks for pointing that out, didn't know.

**Comment by lervag** [comment]
> > texlab is currently described in :help vimtex-nf-lsp, so a good starting point would be to update this text and more precisely describe the overlap.
> 
> The **non-features** wording might be perceived as sending out a stronger negative signal ...

Good point; I agree.

> Instead, I think the section could be titled "Integration with other relevant technologies" or "Interoperation with other plugins" or something which conveys a more inclusive/welcoming tone (reflecting the openness of your response here).

Good suggestion. I think you have good points here. I've pushed an update that I hope you find agreeable; and please, if you have further ideas for improving this, don't hesitate to open a PR!

> In such a documentation section, providing an initial example set of keymaps and configs explaining how to co-operate with texlab shall enrich the user experience, imo.

Agreed, at least in part. I believe the keymaps for texlab are not specific. Instead, you define keymaps for the LSP in general, and these should be valid across languages and LSP servers. But I agree it could be expanded upon. Again, feel free to suggest improvements!

> For visibility/discoverability reasons, would you be okay to indicate that you are willing to accept example LSP configs as documentation PRs.

I want to keep the docs as concise as possible. But I don't mind PRs that improve and expand upon topics in way that make sense (e.g. for visibility and discoverability). I would be OK to indicate this type of willingness - feel free to suggest how in a PR.

> It would be good to indicate this inline in the documentation, but I suspect that might be too deeply embedded in the text to make an impact. The `README.md` itself is a great place to indicate that you are open to LSP-related PRs. Maybe you could also tag this GitHub issue as "help-wanted" etc. etc.

This issue is both old and closed, so I don't think it will help to tag it. Perhaps better to open a new issue with a "feature" request to improve docs with a suggestion on how to integrate VimTeX with texlab. If you both refer to the current text, this particular issue, and add the tag "help-wanted", then it may attract a contribution.

I don't think the README.md is a good place to indicate this, but you could open a PR to show me what you mean. If I agree, I would merge the PR. :)

**Comment by lervag** [comment]
(By the way - this thread is now so long that I'll lock it when we've settled these final points.)

**Comment by krishnakumarg1984** [comment]
Just wanted to let you know that the links to the gists for configuring for the co-existence of texlab and vimtex are now invalid, and give 404 errors.

@clason Would you mind pointing us to updated links for your texlab and vimtex configuration? These could be a good starting point for users who arrive in this issue thread by following vimtex's documentation.

**Comment by clason** [comment]
Those links were never meant to be public; just some snippets I shared privately.

**Comment by lervag** [comment]
> Just wanted to let you know that the links to the gists for configuring for the co-existence of texlab and vimtex are now invalid, and give 404 errors.

Are you referring to something in the docs; i.e., should I update the docs?

**Comment by krishnakumarg1984** [comment]
No. Just clason's links in this thread. The docs mention this thread, and when I tried to follow the gists linked here, they are gone. This could be frustrating for others who similarly come to this thread by following the documentation.
